File: ruby/word-count/word_count.rb
class Phrase
  def initialize(words)
    @words = words
  end

  def word_count
    hash = Hash.new

    # Split words to an array avoiding special characters
    array = @words.gsub(/[^a-zA-Z0-9']|'(?!\w\s)/, ' ').strip.split(%r{[, ]\s*})

    # Normalize case
    array.each { |w| w.downcase! }

    # Count words
    array.map { |w| hash[w] = array.count(w) }
      
    hash
  end
end

File: typescript/word-count/word-count.ts
export default class Words {
  count(phrase: string) {
    const result = new Map();
    phrase.split(new RegExp(" {1,}|\\n|\\t{1,}")).map((word) => {
      if (word) {
        word = word.toLowerCase();
        if (result.has(word)) {
          let value = result.get(word);
          result.set(word, ++value);
        } else {
          result.set(word, 1);
        }
      }
    });
    return result;
  }
}
